### PR TITLE
Change OS.Dir.user and add OS.Dir.expand_tilde

### PR DESCRIPTION
**Pull request by favonia**
## Changes to Os.Dir.user
* On all operating systems, `HOME` is consulted first. (Previously, the `passwd` is consulted first on non-Windows.)
* On Windows, `USERPROFILE` is consulted next.
* The error message on Windows will be about `USERPROFILE`, while the message on other operating systems will be about `HOME`. (I don't feel strongly about this, but it looks reasonable to me.)

## Questions about OS.Dir.expand_tile
* ~~Maybe it should be placed in OS.Path because the path to be expanded can be a path to a file, not a directory? But OS.Dir already depends on OS.Path.~~ Answer: it is fine.

## Status
This PR only partially addresses #77 and #93: I did not implement the mechanism to fetch `FOLDERID_Profile` on Windows as hinted by @dra27 in #77, and I want other utility functions planned in #93.

### CODE REVIEW

**Comment by dbuenzli** [comment]
Hello @favonia sorry for not reacting to your messages in the issue tracker it seems I missed them.

What's the rationale behind introducing `home` as an alias to `user` ? I'd rather not have multiple name for the same functionality.

Regarding tilde expansion I'd rather leave that to shells. But if you want to add it please make it [POSIX compliant](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_01) to avoid surprises and future bug reports (i.e. handle the `~username/` case). Also I would rather call it `expand_tilde` it's clearer (having it in `OS.Dir` is fine).

**Comment by favonia** [comment]
> What's the rationale behind introducing `home` as an alias to `user` ? I'd rather not have multiple name for the same functionality.

Part of the reason to have `home` is to match `expand_home`, and I guess that point is moot if we use `expand_tilde`.

> Regarding tilde expansion I'd rather leave that to shells. But if you want to add it please make it [POSIX compliant](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_01) to avoid surprises and future bug reports (i.e. handle the `~username/` case). Also I would rather call it `expand_tilde` it's clearer (having it in `OS.Dir` is fine).

Sure. I can easily make it POSIX-compliant. I did the limited version only because I failed to convince others when advocating that for Agda library management.

**Comment by dbuenzli** [comment]
> I did the limited version only because I thought others would think the full version is overkill.

Well I don't want to have to handle the bug report `~username/` is not working :-)

File: src/bos.mli [hunk 1]
```diff
@@ -1031,9 +1031,16 @@ contents d >>= Path.fold err dotfiles elements traverse f acc
...
-        with the user id of the process. If this fails or on Windows
-        falls back to parse a path from the [HOME] environment variable. *)
+        the process. Determined by parsing the [HOME] environment variable.
+        If this fails, falls back to parse the [USERPROFILE] variable on Windows
+        or consult the [passwd] database with the user id of the process
+        in other operating systems. *)
+
+    val expand_tilde : Fpath.t -> (Fpath.t, 'e) result
+    (** [expand_tilde p] expands a tilde-prefixed path [p] according to the
+        {{https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_01}POSIX standard},
```
Comment (dbuenzli) [inline src/bos.mli:1041]: you are missing a `:` here.

**Comment by favonia** [comment]
The current code is wrong because it would expand `~` in `C:~`. And I just learned that the internal representation of `Fpath.t` is a string. Will update the code accordingly.

**Comment by dbuenzli** [comment]
Could please a Windows guru like @dra27 or @jonahbeckford confirm the [high-level Windows logic](https://github.com/dbuenzli/bos/pull/96/files#diff-15635f79ea90207203625e70db1cfad147e0e37559e979b73a94364abeada83bR1034-R1037) makes sense (even if imperfect).

**Comment by jonahbeckford** [comment]
TLDR: My opinion: I would only use HOME if the currently executing program is linked to `cygwin1.dll` (the Unix compatibility layer used by Cygwin and MSYS2). Simply checking for the existence of HOME will break some Windows users, and will in fact break Diskuv OCaml Windows users. And since checking for `cygwin1.dll` is difficult to detect, just use USERPROFILE exclusively on Windows.

*The prior code already seems to look at HOME, so `OS.Dir.user` was probably breaking some Windows use cases already.*

In what you see below, `with-dkml` is the wrapper from Diskuv OCaml that Opam is configured to use ... among other things, it delegates Unix commands (`sh`, `env`, etc.) to MSYS2 on Windows.

```powershell
C:\Users\beckf> with-dkml env | Select-String ^HOME=
HOME=/home/beckf

C:\Users\beckf> with-dkml env | Select-String ^USERPROFILE=
USERPROFILE=C:\Users\beckf
```

Now let's say you compile a "native" Windows executable that uses `OS.Dir.user`. By "native" I mean:
* you use the [Cygwin's "MinGW" GCC compiler](https://cygwin.com/packages/summary/mingw64-x86_64-gcc-core.html), the Visual Studio compiler, or a clang compiler
* you do not use [Cygwin's standard GCC compiler](https://cygwin.com/packages/summary/gcc-core.html) (which would link your executable to cygwin1.dll)

Did you notice that `HOME` was a Unix path? It is required to be present so that `cygwin1.dll` linked programs (ex. `sh build.sh`, `rm -rf _build`) can function. But only `cygwin1.dll` knows how to map `/home/beckf` to `C:\Users\beckf`!

The only way a native Windows executable could ever resolve `/home/beckf` is to see if `cygpath.exe` is in the PATH, and then spawn `cygpath.exe -aw /home/beckf`. I highly doubt you want your `bos` users to have to do that.

**Comment by dbuenzli** [comment]
> TLDR: My opinion: I would only use HOME on Windows if the currently executing program is linked to `cygwin1.dll` (the Unix compatibility layer used by Cygwin and MSYS2).

Thanks Jonah! Doesn't that coincide with `Sys.cygwin = true` ? (I'm always extremely confused by all possible setups on the Windows side of things :-)

**Comment by jonahbeckford** [comment]
> Doesn't that coincide with `Sys.cygwin = true` ?

I think that is true. From what I understand from https://github.com/ocaml/ocaml/blob/f150a85e8453a77243d9b7fba4c4459df6745b68/configure#L13370-L13403, `Sys.cygwin = true` means the OCaml runtime was compiled under the Cygwin standard GCC compiler (`*-*-cygwin*`). The Cygwin MinGW GCC compiler (`*-*-mingw32`) should give `false`. And the Visual Studio compiled Diskuv OCaml system gives `false` as well.

 So `Sys.cygwin = true` is a good precondition for checking `HOME` on Windows.

**Comment by favonia** [comment]
@jonahbeckford Thanks for the insight! I don't have a Windows machine so I could only rely on others. To clarify, with Cygwins I believe Sys.os_type will be "Cygwin" (instead of "Windows") and it will use the same logic for other Unix-like systems (which checks HOME and then reads passwd if that fails). So, if I understand your suggestion correctly, when Sys.os_type is "Win32", we should just check USERPROFILE (or even better, querying FOLDERID_Profile).

**Comment by favonia** [comment]
@jonahbeckford @dbuenzli Done, with documentation updated.

**Comment by jonahbeckford** [comment]
Yes, the high-level logic looks right to me.

File: src/bos_os_dir.ml [hunk 2]
```diff
@@ -70,30 +70,80 @@ let fold_contents ?err ?dotfiles ?elements ?traverse f acc d =
...
+                            var_name
   in
-  if Sys.os_type = "Win32" then env_var_fallback () else
-  try
-    let uid = Unix.getuid () in
-    let home = (Unix.getpwuid uid).Unix.pw_dir in
-    match Fpath.of_string home with
+  if Sys.os_type = "Win32" then
+    (* We should check USERPROFILE exclusively (without HOME) because otherwise it could break
+       Diskuv OCaml Windows users. See the discussion at https://github.com/dbuenzli/bos/pull/96.
```
Comment (jonahbeckford) [inline src/bos_os_dir.ml:94]:
```suggestion
       native Windows users. See the discussion at https://github.com/dbuenzli/bos/pull/96.
```

Comment (favonia) [reply src/bos_os_dir.ml]: @jonahbeckford FYI: I changed it to "... break a Windows setup with MSVC++ or Mingw" for consistency.

**Comment by favonia** [comment]
@dbuenzli I got two ideas:

1. I added `user_of` to get the home directory of a user (by its login name).
2. Maybe we could add an OS-independent `expand_tilde` to the fpath library instead? It will take a function of type ``string option -> (t, [`Msg of string]) result`` to get the replacement. Here's the intended usage:
   ```ocaml
   Fpath.expand_tilde (function None -> user () | Some login_name -> user_of login_name) p
   ```
   It would be almost the `expand_tilde_gen` function in this PR, and I think the code can be further improved if we can access the underlying string representation directly. I also found a different usage of this OS-independent expansion in the library I am currently writing.

**Comment by favonia** [comment]
> Not very fond of the idea. Most applications just want the user directory, multi-user environments are rather on the downhill. I prefer what was before, simple and easy to use.

Okay! I turned back the clock.

**Comment by dbuenzli** [comment]
Thanks, your patch is in.
